### ISSUE TITLE
[no-shadow] False positive in function block as variable is initialized

### ISSUE THREAD

**Issue by binoche9** [typescript-eslint/typescript-eslint#2889]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-shadow": "error"
  }
}
```

```TS
const foo = func(() => {
  const foo = 3;
  return foo + 2;
})
```

**Expected Result**

No error should be thrown because inside the function block the outer `foo` hasn't been initialized yet, so it shouldn't be in scope.

**Actual Result**

The lint rule thinks that the inner `foo` is shadowing the outer `foo`.

**Additional Info**

I thought the `{ hoist: "never" }` option may help but it doesn't seem to.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.10.0` |
| `@typescript-eslint/parser`        | `4.10.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.13.1` |

**Comment by bradzacher** [comment]
This is a bug in the underlying rule.
Please file an issue with the eslint repo.

[eslint repl](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXNoYWRvdzogWydlcnJvciddICovXG5cbmNvbnN0IGZvbyA9IGZ1bmMoKCkgPT4ge1xuICBjb25zdCBmb28gPSAzO1xuICByZXR1cm4gZm9vICsgMjtcbn0pOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by binoche9** [comment]
Ah good point. Just to clarify, would fixing it in ESLint automatically update the behavior for TypeScript too, or would we have to update this repo separately (since this repo has its own parser/scope manager)?

**Comment by bradzacher** [comment]
It won't - specifically because our version of no-shadow is _implemented_ as a fork, not as an extension (because it was easier to implement that way).

However as with all of our extension rules - we seek to maintain compatibility with the base rule and report on the same things the base rule reports on - regardless of implementation details.

If they release a fixed version, we can implement the same change here.
